lib/active_merchant/billing/gateways/adyen.rb

```diff
@@ -622,20 +622,31 @@ def add_mpi_data_for_network_tokenization_card(post, payment, options)
 
         post[:mpiData] = {}
         post[:mpiData][:authenticationResponse] = 'Y'
-        post[:mpiData][:cavv] = payment.payment_cryptogram
+        if NETWORK_TOKENIZATION_CARD_SOURCE[payment.source.to_s].nil? && options[:switch_cryptogram_mapping_nt]
+          post[:mpiData][:tokenAuthenticationVerificationValue] = payment.payment_cryptogram
+        else
+          post[:mpiData][:cavv] = payment.payment_cryptogram
+        end
         post[:mpiData][:directoryResponse] = 'Y'
         post[:mpiData][:eci] = payment.eci || '07'
       end
 
-      def add_recurring_contract(post, options = {})
-        return unless options[:recurring_contract_type]
+      def add_recurring_contract(post, options = {}, payment = nil)
+        return unless options[:recurring_contract_type] || (payment.try(:source) && options[:switch_cryptogram_mapping_nt])
 
-        post[:recurring] = {}
-        post[:recurring][:contract] = options[:recurring_contract_type]
+        post[:recurring] ||= {}
+        post[:recurring][:contract] = options[:recurring_contract_type] if options[:recurring_contract_type]
         post[:recurring][:recurringDetailName] = options[:recurring_detail_name] if options[:recurring_detail_name]
         post[:recurring][:recurringExpiry] = options[:recurring_expiry] if options[:recurring_expiry]
         post[:recurring][:recurringFrequency] = options[:recurring_frequency] if options[:recurring_frequency]
         post[:recurring][:tokenService] = options[:token_service] if options[:token_service]
+        if payment.try(:source) && options[:switch_cryptogram_mapping_nt]
```

Comment:
This needs to be `payment.try(:source) == :network_token && options[:switch_cryptogram_mapping_nt]`

Reply:
Once these are fixed I'll pull my needs updates tag off.

Reply:
@DustinHaefele, @almalee24 In case of googlepay  and applepay what would happen in 643 to 648?
for network tokens the `post[:recurring][:contract]` & `post[:recurring][:tokenService]` would be missed in both cases and according to my understanding of the info in https://spreedly.slack.com/archives/CH7B6NWA0/p1709310814097569 those are required.
Please check and let me know what you think

Reply:
For googlepay and applepay we don't want the values set in 643-648. This is because lines 643-648 are for true network_tokens that are issued by visa and mastercard. In the case of applepay and googlepay the customer would have to pass in the `options[:token_service]` and `options[:recurring_contract_type]` values if they wanted to add them here.

Reply:
Understood, fix added. Thanks!